context/Web3Context.tsx

```diff
@@ -0,0 +1,71 @@
+"use client";
+
+import { ModalId, useModalContext } from "@/context/ModalContext";
+import { PropsWithChildren, createContext, useContext, useState } from "react";
+import { waitForTransactionReceipt } from "@wagmi/core";
+import { type WriteContractParameters, type Config } from "@wagmi/core";
+
+import { writeContract as wagmiWriteContract } from "wagmi/actions";
+import { TransactionModal } from "@/app/components";
+
+export interface Web3ContextProps {
+  writeContractShowModal: (
+    config: Config,
+    args: WriteContractParameters
+  ) => Promise<string | undefined>;
+}
+
+export const Web3Context = createContext<Web3ContextProps>({
+  writeContractShowModal: async (config, args) => "",
+});
+
+export const Web3Provider = ({ children }: PropsWithChildren) => {
+  const [txHash, setTxHash] = useState<string>("");
+  const [isTxLoading, setIsTxLoading] = useState(false);
+  const [isError, setIsError] = useState(false);
+
+  const { openModal } = useModalContext();
+
+  const writeContractShowModal = async (
+    config: Config,
+    args: WriteContractParameters
+  ): Promise<string | undefined> => {
+    setIsTxLoading(true);
+
+    try {
+      const txHash = await wagmiWriteContract(config, args);
+      setTxHash(txHash);
+
+      openModal(ModalId.WAITING_TRANSACTION);
+
+      await waitForTransactionReceipt(config, {
+        hash: txHash,
+      });
+
+      return txHash;
+    } catch (error) {
+      setIsError(true);
+      console.error(error);
+      return undefined;
```

Comment:
It's better to throw an error here.

Reply:
Or maybe do nothing has we are dealing with the error already.

Reply:
I think we need to throw an error, so we have to catch it. By catching we prevent calling `onApprove` if something goes wrong, for example. https://github.com/SwaprHQ/presagio/pull/39/files#diff-c24360049586ecf1a45d448ed999a5f309471ae6ca58880e97df385e7a6da1abR87